src/finiteVolume/processes/les.cpp

```diff
@@ -0,0 +1,240 @@
+
+    // Compute the LES stress tensor tau
+    PetscReal LesTau[9];  // Maximum size without symmetry
+    const int EV_FIELD = 0;
+    PetscReal mut;
+
+    ierr = LesViscosity(dim, fg, field, uOff, mut);
+    CHKERRQ(ierr);
+    ierr = CompressibleFlowComputeLesStressTensor(dim, fg, gradAux + aOff_x[VEL], uOff, field, ctx, LesTau);
+    CHKERRQ(ierr);
+
+    const PetscReal areaMag = utilities::MathUtilities::MagVector(dim, fg->normal);
+
+    // energy equation
+    for (PetscInt ev = 0; ev < flowParameters->numberEV; ++ev) {
+        for (PetscInt d = 0; d < dim; ++d) {
+            PetscReal LesEvFlux = 0.0;
+            for (PetscInt c = 0; c < dim; ++c) {
+                //
+                LesEvFlux += -areaMag * sqrt(areaMag) * density * LesTau[d * dim + c] * LesTau[d * dim + c] / mut;
+            }
+
+            //  LESevFlux( rho Di dEVi/dx + rho Di dEVi/dy + rho Di dEVi//dz) . n A +  LESevFlux(-rho ce EV^3/2 ) . n A
+            const int offset = aOff_x[EV_FIELD] + d;
+            LesEvFlux += -fg->normal[d] * (density * mut * gradAux[offset] - c_e * field[uOff[EV_FIELD]] * sqrt(field[uOff[EV_FIELD]] / density));
+
+            flux[ev] = LesEvFlux;
+        }
+    }
+
+    PetscFunctionReturn(0);
+}
+
+PetscErrorCode ablate::finiteVolume::processes::LES::CompressibleFlowComputeLesStressTensor(PetscInt dim, const PetscFVFaceGeom* fg, const PetscReal* gradVel, const PetscInt uOff[],
+                                                                                            const PetscScalar field[], void* ctx, PetscReal* LesTau) {
+    PetscFunctionBeginUser;
+    // pre-compute the div of the velocity field
+    PetscReal divVel = 0.0;
+    PetscReal mut;
+
+    PetscErrorCode ierr;
+    ierr = LesViscosity(dim, fg, field, uOff, mut);
+    CHKERRQ(ierr);
+
+    for (PetscInt c = 0; c < dim; ++c) {
+        divVel += gradVel[c * dim + c];
+    }
+
+    // March over each velocity component, u, v, w
+    for (PetscInt c = 0; c < dim; ++c) {
+        // March over each physical coordinates
+        for (PetscInt d = 0; d < dim; ++d) {
+            if (d == c) {
+                // for the xx, yy, zz, components
+                LesTau[c * dim + d] = 2.0 * mut * ((gradVel[c * dim + d]) - divVel / 3.0);
+            } else {
+                // for xy, xz, etc
+                LesTau[c * dim + d] = mut * ((gradVel[c * dim + d]) + (gradVel[d * dim + c]));
+            }
+        }
+    }
+    PetscFunctionReturn(0);
+}
+
+PetscErrorCode ablate::finiteVolume::processes::LES::LesViscosity(PetscInt dim, const PetscFVFaceGeom* fg, const PetscScalar field[], const PetscInt uOff[], PetscReal& mut) {
+    PetscFunctionBeginUser;
+    const int euler = 0;
+    const int EV_FIELD = 0;
+
+    const PetscReal areaMag = utilities::MathUtilities::MagVector(dim, fg->normal);
+
+    // get the current density from euler
+    const PetscReal density = field[uOff[euler] + CompressibleFlowFields::RHO];
+
+    for (PetscInt c = 0; c < dim; ++c) {
```

Comment:
Why are you marching over the number of dimensions?  Where component in the ev field is tke?